qualtran/bloqs/mod_arithmetic/mod_subtraction.py

```diff
@@ -247,3 +249,158 @@ def _cmod_neg() -> CModNeg:
+    n, p = sympy.symbols('n p')
+    symb_modsub = ModSub(QUInt(n), p)
+    return symb_modsub
+
+
+_MOD_SUB_DOC = BloqDocSpec(bloq_cls=ModSub, examples=[_symb_modsub])
+
+
+@frozen
+class CModSub(Bloq):
+    r"""Performs controlled modular subtraction.
+
+    Applies the operation $\ket{c}\ket{x} \ket{y} \rightarrow \ket{c}\ket{x} \ket{y-cx \% p}$
+
+    Args:
+        dtype: Datatype of the register.
+        p: The modulus for the negation.
+        cv: value at which the bloq is active.
+
+    Registers:
+        ctrl: control register.
+        x: The register contraining the first integer.
+        y: The register contraining the second integer.
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585)
+        Fig 6b and 8
+    """
+
+    dtype: Union[QUInt, QMontgomeryUInt]
+    mod: 'SymbolicInt'
+    cv: int = 1
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [Register('ctrl', QBit()), Register('x', self.dtype), Register('y', self.dtype)]
+        )
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', ctrl: Soquet, x: Soquet, y: Soquet
+    ) -> Dict[str, 'SoquetT']:
+        x = bb.add(BitwiseNot(self.dtype), x=x)
+        x = bb.add(AddK(self.dtype.bitsize, self.mod + 1, signed=False), x=x)
+        ctrl, x, y = bb.add(CModAdd(self.dtype, self.mod, self.cv), ctrl=ctrl, x=x, y=y)
+        x = bb.add(AddK(self.dtype.bitsize, self.mod + 1, signed=False).adjoint(), x=x)
+        x = bb.add(BitwiseNot(self.dtype), x=x)
+        return {'ctrl': ctrl, 'x': x, 'y': y}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {
+            (BitwiseNot(self.dtype), 2),
+            (AddK(self.dtype.bitsize, self.mod + 1, signed=False), 1),
+            (AddK(self.dtype.bitsize, self.mod + 1, signed=False).adjoint(), 1),
+            (CModAdd(self.dtype, self.mod, self.cv), 1),
+        }
+
+    def wire_symbol(
+        self, reg: Optional['Register'], idx: Tuple[int, ...] = tuple()
+    ) -> 'WireSymbol':
+        if reg is None:
+            return Text("")
+        if reg.name == 'ctrl':
+            return Circle(filled=self.cv == 1)
+        if reg.name == 'x':
+            return TextBox('$x$')
+        if reg.name == 'y':
+            return TextBox('$y-x$')
+        raise ValueError(f'Unrecognized register name {reg.name}')
+
+    def on_classical_vals(
+        self, ctrl: 'ClassicalValT', x: 'ClassicalValT', y: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        if ctrl == self.cv and 0 <= x < self.mod and 0 <= y < self.mod:
+            y -= x
+            if y < 0:
+                y += self.mod
+        return {'ctrl': ctrl, 'x': x, 'y': y}
+
+
+@bloq_example
+def _symb_cmodsub() -> CModSub:
+    n, p = sympy.symbols('n p')
+    symb_cmodsub = CModSub(QUInt(n), p)
+    return symb_cmodsub
```

Comment:
we typically name these `bloq_variant` so `cmodsub_symb` (not `symb_cmodsub` like present). If you change this -- it might be easier to delete the existing notebook before regenerating it or the cells will be out of order, sorry

Reply:
done